### PR TITLE
Replace the global lock by the SubnetSet lock for Subnet creation

### PR DESCRIPTION
**Pull request by yanjunz97**

### CODE REVIEW

File: pkg/controllers/common/utils.go [hunk 1]
```diff
@@ -240,3 +240,17 @@ func NewStatusUpdater(client k8sclient.Client, nsxConfig *config.NSXOperatorConf
 		ResourceType:    resourceType,
 	}
 }
+
+func lockSubnetSet(uuid types.UID) {
+	lock := sync.Mutex{}
+	subnetLock, _ := SubnetSetLocks.LoadOrStore(uuid, &lock)
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils.go:246]: Not quite get the question. This function is only for lock. When calling this function in AllocateSubnetFromSubnetSet I call defer unlock:
https://github.com/vmware-tanzu/nsx-operator/blob/5f79a21bcafa0b29be57671974bffbb2ff7b48bd/pkg/controllers/common/utils.go#L33

[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils.go:246]: Not quite get the question. This function is only for lock. When calling this function in AllocateSubnetFromSubnetSet I call defer unlock:
https://github.com/vmware-tanzu/nsx-operator/blob/5f79a21bcafa0b29be57671974bffbb2ff7b48bd/pkg/controllers/common/utils.go#L33

File: pkg/controllers/subnetset/subnetset_controller.go [hunk 2]
```diff
@@ -260,6 +261,15 @@ func (r *SubnetSetReconciler) CollectGarbage(ctx context.Context) {
 			r.StatusUpdater.IncreaseDeleteSuccessTotal()
 		}
 	}
+
+	// clean the SubnetSet lock used to create Subnet
+	common.SubnetSetLocks.Range(func(key, value interface{}) bool {
+		uuid := key.(types.UID)
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/subnetset/subnetset_controller.go:267]: Here I delete the subnetset lock if the corresponding subnetset CR does not exist. In this subnetset gc, we also delete the subnet if the corresponding subnetset CR does not exist above this part of code. So currently the deletion of the key is after the subnet deletion.
https://github.com/vmware-tanzu/nsx-operator/blob/5f79a21bcafa0b29be57671974bffbb2ff7b48bd/pkg/controllers/subnetset/subnetset_controller.go#L253-L256

File: pkg/controllers/common/utils.go [hunk 3]
```diff
@@ -240,3 +240,17 @@ func NewStatusUpdater(client k8sclient.Client, nsxConfig *config.NSXOperatorConf
...
+	log.V(1).Info("Lock SubnetSet", "uuid", uuid)
+	subnetLock.(*sync.Mutex).Lock()
+}
+
+func unlockSubnetSet(uuid types.UID) {
+	if subnetLock, existed := SubnetSetLocks.Load(uuid); existed {
```
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils.go:252]: I do not use lock to sync the SubnetSetLocks. I think the sync.Map is thread safe and no extra lock is required.

Would you mean there is a case that we get existed=false here and skip the unlock but after that we call lockSubnetSet which may cause a deadlock? But in our use case unlock is only called after lock, so it wont happen?

File: pkg/controllers/common/utils.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yanjunz97) [reply pkg/controllers/common/utils.go]: Fix a naming typo, hope that does not confusing you too much...

**Review by TaoZou1 — DISMISSED** [review]

**Review by wenqiq — COMMENTED** [review]
LGTM and just one nit. The new lines in `pkg/controllers/subnetset/subnetset_controller.go` are not covered by unit-test.

**Review by wenqiq — APPROVED** [review]
LGTM

**Review by TaoZou1 — APPROVED** [review]

**Pull request merged**